### PR TITLE
[dpdk] Update to version v22.03

### PR DESCRIPTION
**Pull request by ljishen**
**Describe the pull request**

- #### What does your PR fix?
  Update dpdk to version v22.03

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  I've removed the `x64-linux=fail` from the file to mark that this port is supported by that Linux triplet.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  `Yes`

### CODE REVIEW

**Comment by dg0yt** [comment]
I think both the old and the new attempt to provide a CMake config are full of problems. In addition, it is not from upstream, so it is should be in the `unofficial` namespace.

**Comment by Cheney-W** [comment]
@ljishen Could you please handle the failure in CI test? [Failure log](https://dev.azure.com/vcpkg/_apis/resources/Containers/17241369/failure%20logs%20for%20x64-linux?itemPath=failure%20logs%20for%20x64-linux%2Fdpdk&%24format=zip&saveAbsolutePath=false)

**Comment by ljishen** [comment]
> @ljishen Could you please handle the failure in CI test? [Failure log](https://dev.azure.com/vcpkg/_apis/resources/Containers/17241369/failure%20logs%20for%20x64-linux?itemPath=failure%20logs%20for%20x64-linux%2Fdpdk&%24format=zip&saveAbsolutePath=false)

The error is expected currently without the #25372. I've figured out how to separate static from dynamic build in dpdk according to @dg0yt's requirement. There is one more issue I need to fix before I can update this PR. Sorry about that.

**Comment by ljishen** [comment]
Is there a way to view the error log?

**Comment by Cheney-W** [comment]
Could you please open this: https://dev.azure.com/vcpkg/public/_build/results?buildId=74149&view=artifacts&pathAsName=false&type=publishedArtifacts ? 
All failure logs are stored there.

**Comment by ljishen** [comment]
> Could you please open this: https://dev.azure.com/vcpkg/public/_build/results?buildId=74149&view=artifacts&pathAsName=false&type=publishedArtifacts ? All failure logs are stored there.

Thanks so much, @Cheney-W. This PR is ready to be reviewed.

**Comment by Cheney-W** [comment]
Please remove the record about dpdk from `ci.baseline.txt`, because you have added `"supports": "linux | freebsd"` into the vcpkg.json file. Thank you!

**Comment by ljishen** [comment]
@Cheney-W I've removed them. Please take a look. Thanks!

**Comment by ljishen** [comment]
@Cheney-W Is this PR ready to be merged? Thanks.

**Comment by Cheney-W** [comment]
Yes

**Review by Thomas1664 — CHANGES_REQUESTED** [review]

File: ports/dpdk/portfile.cmake [hunk 1]
```diff
@@ -1,43 +1,95 @@
...
+  "${CMAKE_CURRENT_LIST_DIR}/unofficial-${PORT}-config.cmake.in"
+  "${CURRENT_PACKAGES_DIR}/share/unofficial-${PORT}/unofficial-${PORT}-config.cmake"
+  @ONLY)
+
+file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage"
+     DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+file(
+  INSTALL "${SOURCE_PATH}/license/README"
+  DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
+  RENAME copyright)
```
[the remark this answers is not in the document]
Comment (ljishen) [reply ports/dpdk/portfile.cmake:88]: What should I do currently with this?

File: ports/dpdk/portfile.cmake [hunk 2]
```diff
@@ -1,43 +1,95 @@
...
-FILE(INSTALL ${SOURCE_PATH}/Release/lib/ DESTINATION ${CURRENT_PACKAGES_DIR}/lib)
-FILE(INSTALL ${SOURCE_PATH}/Debug/lib/ DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib)
-FILE(INSTALL ${REAL_FILES} DESTINATION ${CURRENT_PACKAGES_DIR}/include/${PORT})
-FILE(INSTALL ${CMAKE_CURRENT_LIST_DIR}/dpdkConfig.cmake DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT})
-FILE(INSTALL ${CMAKE_CURRENT_LIST_DIR}/usage DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT})
-FILE(INSTALL ${SOURCE_PATH}/license/README DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
+if(NOT VCPKG_CMAKE_SYSTEM_NAME OR NOT (VCPKG_CMAKE_SYSTEM_NAME STREQUAL "Linux" OR VCPKG_CMAKE_SYSTEM_NAME STREQUAL "FreeBSD"))
+  message(FATAL_ERROR "Intel DPDK currently only supports Linux/FreeBSD platforms")
+endif()
+
```
[the remark this answers is not in the document]
Comment (ljishen) [reply ports/dpdk/portfile.cmake:4]: OK

File: ports/dpdk/portfile.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (Thomas1664) [reply ports/dpdk/portfile.cmake:84]: Wait until #25239 gets merged. The PR is almoast ready so it should't take long

File: ports/dpdk/vcpkg.json [hunk 3]
```diff
@@ -1,7 +1,41 @@
 {
   "name": "dpdk",
-  "version-string": "19.02",
-  "port-version": 1,
+  "version-string": "22.03",
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/dpdk/vcpkg.json:3]: Yes. It is not the first port which is hit by this.

File: ports/dpdk/portfile.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (ljishen) [reply ports/dpdk/portfile.cmake]: I've updated the port to use the function to be merged. I will rebase it once the code is actually there.

[the remark this answers is not in the document]
Comment (ljishen) [reply ports/dpdk/portfile.cmake]: @Thomas1664 Could I know when will #25239 get merged?

File: ports/dpdk/portfile.cmake [hunk 4]
```diff
@@ -1,43 +1,93 @@
...
+endif()
+x_vcpkg_get_python_packages(PYTHON_VERSION "3" PACKAGES ${PYTHON_PACKAGES})
+
+# Backup the environment variables for pkg-config before calling
+# vcpkg_configure_meson(). Relevant issue:
+# https://github.com/microsoft/vcpkg/pull/25118
+vcpkg_backup_env_variables(VARS PKG_CONFIG PKG_CONFIG_PATH)
+vcpkg_configure_meson(SOURCE_PATH ${SOURCE_PATH} OPTIONS
+                      -Ddisable_drivers=regex/cn9k -Dexamples= ${DPDK_OPTIONS})
+vcpkg_restore_env_variables(VARS PKG_CONFIG PKG_CONFIG_PATH)
```
Comment (dg0yt) [inline ports/dpdk/portfile.cmake:64]: `z_vcpkg_setup/restore_pkgconfig_path` should now work as expected (#25361). Please merge master and test without the explicit backup/restore.

File: ports/dpdk/portfile.cmake [hunk 5]
```diff
@@ -1,43 +1,93 @@
...
+  "${CURRENT_PACKAGES_DIR}/share/unofficial-${PORT}/unofficial-${PORT}-config.cmake"
+  @ONLY)
+
+file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage"
+     DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+
+# Concatenate multiple copyright files used by DPDK to a single `copyright` file
+file(GLOB license_files "${SOURCE_PATH}/license/*.txt")
+file(READ "${SOURCE_PATH}/license/README" comment)
+vcpkg_install_copyright(FILE_LIST ${license_files} COMMENT "${comment}")
```
Comment (dg0yt) [inline ports/dpdk/portfile.cmake:85]: This function shouldn't be used before https://github.com/microsoft/vcpkg/pull/25239 is merged.

**Review by dg0yt — CHANGES_REQUESTED** [review]

**Comment by Cheney-W** [comment]
@ljishen Could you please handle the suggestions of dg0yt? 
Also, I don't think you should let a PR that hasn't been merged block your PR.

**Comment by ljishen** [comment]
> @ljishen Could you please handle the suggestions of dg0yt? Also, I don't think you should let a PR that hasn't been merged block your PR.

Great. Thanks for your support! I'll fix the code.

File: ports/dpdk/portfile.cmake (hunk 5 above)
Comment (ljishen) [reply ports/dpdk/portfile.cmake]: OK

File: ports/dpdk/portfile.cmake (hunk 4 above)
Comment (ljishen) [reply ports/dpdk/portfile.cmake]: Yes. The variable backup is working. Thanks.

**Comment by ljishen** [comment]
@dan-shaw Could you help to merge this one, please?

**Review by Thomas1664 — APPROVED** [review]

**Comment by autoantwort** [comment]
This package contains the following absolute path in `include/rte_build_config.h`:
```c++
#define RTE_EAL_PMD_PATH "/home/leander/vcpkg/packages/dpdk_x64-linux/lib/dpdk/pmds-22.1"
```
and it seems that this path is used in the code. The Problem is not existing when the package is restored from a binary cache. How should we handle this @ljishen?

**Comment by ljishen** [comment]
@[autoantwort](https://github.com/autoantwort) With the linux triplet (e.g., [x64-linux](https://github.com/microsoft/vcpkg/blob/master/triplets/x64-linux.cmake)), this port will only build static libraries. The absolute path you asked contains dpdk's shared libraries and should be empty based on the triplet's requirement. Can you elaborate on the problem you have with that macro definition in the code? Did you have a customized triplet to use dpdk as a shared library?

**Comment by dg0yt** [comment]
Since 10 hours, there is now an x64-linux-dynamic.cmake community triplet.
And in general, there must be no absolute paths in vcpkg packages. They must be relocatable.

**Comment by ljishen** [comment]
OK. I'll update the dpdk port to fix the absolute part problem.

**Comment by autoantwort** [comment]
@ljishen Is there already a PR to fix the absolute path problem? :)

**Pull request merged**